rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,196 @@
+
+Access Requests in the web UI used a client side paginated table, and would fetch every request in the initial API call.
+```go
+// example handler
+func AccessRequestHandler() ([]AccessRequest, error) {
+    accessRequests := []AccessRequest
+    return accessRequests,nil
+}
+```
+ Some users had too many access requests to send over the fetch and the page would break, becoming unusable for larger users. The solution was to server side paginate the Access Requests request. The return used to just be a list of Access Requests, and it has then changed to an object with a list of Access Requests as a field and a nextKey field. 
+ 
+ ```go
+// example handler
+func AccessRequestHandler() ([]AccessRequest, error) {
+    accessRequests := []AccessRequest
+    nextKey := "hi_im_next"
+    return AccessRequestResponse{
+        requests: accessRequests,
+        nextKey: nextKey,
+    }, nil
+}
+```
+
+This breaks backward compatibility because a web client of version n-1 would be expecting a list of requests only, and a web client of n hitting a proxy of n-1 would be expecting to receive the paginated list.
+
+The solution _in this particular case_ was
+1. on the proxy side, check which version of the web client was making the request (by the existence of a specific query param, `limit`) and decide which shape to send
+2. on the web client side, if receiving a response from an old proxy, fallback to the client pagination.
+
+This may work, but it highlights a couple things.
+- we can send any response shape we want with no errors/warnings. 
+- if backward compatibility wasn't forefront of the mind, there was nothing stopping us from just pushing the new code with a break
+- the backend has to check the request shape and the frontend has to check the response shape just to know what type it is receiving.
+
+Using a standard guideline for an rpc service should help us evolve our API overtime in a backward compatible way.
+
+## Solution
+
+**Summary**: Leverage ConnectRPC and create a web service that can use a proto schema to define our requests and response types. Also, optionally, use a modern fetching framework to simplify and unify the way we make api requests in the web client. This will remove boilerplate that we need to use our APIs in the web client, provide type safety out of the box for both server and client code, and provide backward compatibility guardrails when creating/updating our APIs. 
+
+
+### Creating a new rpc service with Connect
+(buf's Connectrpc, not Teleport Connect)
+
+Connect is a library used in making gRPC compatible HTTP apis. It supports [multiple protocols](https://connectrpc.com/docs/introduction#seamless-multi-protocol-support)  and can generate go service code and typescript client code from the proto schema. We do something similar in Teleport Connect where we generate a client for `tsh` to make requests to the server and use the generated typescript code. This would be doing the same for the web client now.
```

Comment:
Support for multiple protocols sounds great, but can we disable unused ones to narrow down attack vectors?

Reply:
Yes you can select your transport.